File: 452.minimum-number-of-arrows-to-burst-balloons.py
#
# @lc app=leetcode id=452 lang=python3
#
# [452] Minimum Number of Arrows to Burst Balloons
#


# @lc code=start
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        # interval schedling problem to find non-overlapping intervals
        # for overlapping intervals, we can ignore it since we can clear those balloons by shooting other balloons
        sorted_points = sorted(points, key=lambda x: x[1])
        intervals = [sorted_points[0]]
        for p in sorted_points[1:]:
            if p[0] <= intervals[-1][1]:
                continue
            else:
                intervals.append(p)
        return len(intervals)


# @lc code=end

File: 452.minimum-number-of-arrows-to-burst-balloons.go
/*
 * @lc app=leetcode id=452 lang=golang
 *
 * [452] Minimum Number of Arrows to Burst Balloons
 */

// @lc code=start
package main

import (
	"sort"
)

func findMinArrowShots(points [][]int) int {
	/*
			   schedule problem: greedy
			   1. get the minimum set of non-overlapped intervals
			   2. the answer if the length of the set since overlapped ballons can be removed when removing ballons in non-interval set
		    Note that the def of "overlap"
	*/
	n := len(points)
	sortedPoints := make([][]int, n)
	copy(sortedPoints, points)
	sort.Slice(sortedPoints, func(i, j int) bool {
		return sortedPoints[i][1] < sortedPoints[j][1]
	})

	numIntervals := 1
	prevIdx := 0
	for i := 1; i < n; i++ {
		if sortedPoints[i][0] > sortedPoints[prevIdx][1] {
			numIntervals++
			prevIdx = i
		}
	}
	return numIntervals
}

// @lc code=end
